dummyfs/dummyfs.c

```diff
@@ -511,470 +635,509 @@ int dummyfs_create(void *ctx, oid_t *dir, const char *name, oid_t *oid, unsigned
-	addr = mmap(NULL, (size + (_PAGE_SIZE - 1)) & ~(_PAGE_SIZE - 1), 0x1 | 0x2, 0, OID_PHYSMEM, (addr_t)addr);
-	if (addr == MAP_FAILED) {
-		free(chunk);
-		object_put(fs, o);
-		dummyfs_destroy(ctx, oid);
-		return -ENOMEM;
+	dummyfs_object_t *o = dummyfs_object_get(fs, oid);
+	if (o == NULL) {
+		_dummyfs_destroy(fs, oid);
+		mutexUnlock(fs->mutex);
+		return -EINVAL;
 	}
-#endif
-
-	chunk->offs = 0;
-	chunk->size = size;
-	chunk->used = size;
-	chunk->data = (void *)((uintptr_t)addr & ~(_PAGE_SIZE - 1)) + ((uintptr_t)addr & (_PAGE_SIZE - 1));
-	chunk->next = chunk;
-	chunk->prev = chunk;
 
-	object_lock(fs, o);
+	o->data = mmap(NULL, (size + (_PAGE_SIZE - 1)) & ~(_PAGE_SIZE - 1), PROT_READ | PROT_WRITE, 0, OID_PHYSMEM, (addr_t)addr);
 	o->size = size;
 	o->mode = OBJECT_MODE_MEM;
-	o->chunks = chunk;
-	object_unlock(fs, o);
-	object_put(fs, o);
 
-	return EOK;
+	dummyfs_object_put(fs, o);
+	mutexUnlock(fs->mutex);
+
+	return 0;
 }
 
 
 int dummyfs_statfs(void *ctx, void *buf, size_t len)
 {
+	TRACE;
 	dummyfs_t *fs = ctx;
 	struct statvfs *st = buf;
 
 	if ((st == NULL) || (len != sizeof(*st))) {
 		return -EINVAL;
 	}
 
-	/* TODO: fs->size access should be protected with a lock */
+	mutexLock(fs->mutex);
 	st->f_bsize = st->f_frsize = 1;
 	st->f_blocks = DUMMYFS_SIZE_MAX;
 	st->f_bavail = st->f_bfree = st->f_blocks - fs->size;
 	st->f_files = 0;
 	st->f_ffree = 0;
 	st->f_favail = 0;
 	st->f_fsid = (unsigned long)fs; /* TODO: filesystem ID should be generated at mount time */
-	st->f_flag = 0;                 /* TODO: mount options should be saved at mount time */
+	st->f_flag = fs->mode;
 	st->f_namemax = 255;            /* TODO: define max filename limit, use 255 limit for now */
+	mutexUnlock(fs->mutex);
 
-	return EOK;
+	return 0;
 }
 
 
 static int dummyfs_alloc(dummyfs_t *fs, const char *data, unsigned long mode)
 {
-	dummyfs_object_t *o;
-	oid_t rootdir;
-	void *mountPoint;
-
-	if (mutexCreate(&fs->mutex) != EOK) {
+	TRACE;
+	if (mutexCreate(&fs->mutex) != 0) {
 		fs->mutex = 0;
 		return -ENOMEM;
 	}
 
-	if (object_init(fs) != EOK) {
-		return -ENOMEM;
-	}
-
-	if (dev_init(fs) != EOK) {
+	if (dummyfs_object_init(fs) != 0) {
 		return -ENOMEM;
 	}
 
 	/* Create root directory */
-	if ((o = object_create(fs)) == NULL) {
+	dummyfs_object_t *o = dummyfs_object_create(fs);
+	if (o == NULL) {
 		return -ENOMEM;
 	}
 
 	o->oid.port = fs->port;
-	o->mode = S_IFDIR | 0666;
+	o->oid.id = DUMMYFS_ROOTID;
+	o->mode = S_IFDIR | DEFFILEMODE;
+	dummyfs_dir_init(fs, o);
 
-	rootdir = o->oid;
+	oid_t rootdir = o->oid;
 	fs->parent = rootdir;
+	fs->mode = mode;
 
 	if (data != NULL) {
 		fs->mountpt = resolve_path(data, NULL, 1, 0);
 		if (fs->mountpt == NULL) {
 			return -ENOMEM;
 		}
 
-		if (lookup(fs->mountpt, &fs->parent, NULL) < EOK) {
+		size_t mntlen = strlen(fs->mountpt);
+		char *parent = malloc(mntlen + 3 + 1);
+		if (parent == NULL) {
+			return -ENOMEM;
+		}
+		memcpy(parent, fs->mountpt, mntlen + 1);
+		strcat(parent, "/..");
```

Comment:
without `strlen(parent)` in `strcat()`
```suggestion
		strcpy(parent + mntlen, "/..");
```

Reply:
Applied, thanks!